<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/04_window_5min_episodes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB04 — Detecção de Episódios Críticos

## 1. Contexto

A partir da série temporal contínua construída no NB03, este notebook realiza a
identificação de janelas críticas e sua consolidação em episódios críticos.

A análise deixa de considerar eventos isolados e passa a focar em padrões agregados ao
longo do tempo, permitindo identificar períodos de degradação operacional com base em
uma métrica relativa de falha por janela.

Como a execução atual do NB03 preservou o início da série e confirmou a presença de
`bucket_id = 0`, esta etapa também passa a documentar explicitamente o comportamento
do bucket inicial e sua eventual influência sobre os parâmetros estatísticos globais.

## 2. Objetivo

O objetivo deste notebook é:
- definir uma métrica representativa de falha por janela;
- estabelecer um critério estatístico para identificação de criticidade;
- detectar janelas críticas na série temporal contínua;
- consolidar janelas críticas contíguas em episódios críticos;
- registrar diagnósticos que permitam avaliar a sensibilidade dos parâmetros globais,
  especialmente em relação ao bucket inicial da série.

## 3. Papel no Pipeline Atualizado

Este notebook desempenha papel central na estruturação semântica do pipeline.

Seu papel é transformar a série agregada em uma representação segmentada composta por:
- janelas não críticas;
- janelas críticas;
- episódios críticos delimitados por contiguidade temporal.

Essa estrutura fornece a referência a partir da qual, nas etapas posteriores, serão
definidos estados operacionais, construídos atributos supervisionados e formulado o
problema preditivo.

## 4. Métrica de Detecção

A métrica adotada para caracterizar criticidade em cada janela é:

`fail_rate = n_failed / n_events`

Essa escolha permite capturar a intensidade relativa de falhas, reduzindo viés associado
ao volume absoluto de eventos.

Para janelas com `n_events = 0`, define-se `fail_rate = 0.0`, preservando a continuidade
da série temporal sem introduzir valores indefinidos no fluxo principal de análise.

## 5. Critério de Criticidade

A identificação oficial de janelas críticas é realizada por meio de um limiar estatístico
global baseado na distribuição da taxa de falha:

`threshold = μ + 2σ`

onde:
- `μ` representa a média global de `fail_rate`;
- `σ` representa o desvio-padrão global de `fail_rate`.

Uma janela é classificada como crítica quando:

`fail_rate >= threshold`

Além da regra oficial, este notebook registra diagnósticos complementares para avaliar o
efeito de decisões metodológicas e características específicas da série, sem alterar, nesta
etapa, a definição oficial utilizada no pipeline.

## 6. Construção de Episódios

As janelas classificadas como críticas são agrupadas em intervalos contíguos, formando
episódios críticos.

Cada episódio representa um período contínuo de comportamento anômalo ou degradação
operacional, sendo caracterizado por:
- início e fim;
- duração em janelas;
- duração em minutos;
- intensidade máxima e média da taxa de falha;
- estatísticas auxiliares de volume absoluto.

## 7. Saídas Esperadas

Ao final deste notebook, espera-se obter:
- identificação das janelas críticas;
- consolidação de episódios críticos;
- dataset estruturado contendo informações dos episódios;
- resumo quantitativo dos parâmetros estatísticos adotados;
- diagnósticos sobre o bucket inicial e sobre a robustez descritiva da segmentação.

## 8. Relação com as Próximas Etapas

Este notebook alimenta diretamente:
- NB05 — Engenharia de atributos;
- NB06 — Derivação de estados e formulação supervisionada;
- etapas posteriores do pipeline que dependem da delimitação consistente dos episódios.

A qualidade da detecção de episódios impacta diretamente a definição dos estados
operacionais e dos rótulos utilizados na modelagem supervisionada.

## 9. Observações Metodológicas

A métrica baseada em taxa reduz viés associado ao volume bruto de eventos, mas a
segmentação resultante continua sensível à estrutura global da série.

Por esse motivo, esta etapa deve documentar:
- o número de janelas vazias;
- a presença e o comportamento do bucket inicial;
- a eventual influência desse bucket sobre `μ`, `σ` e `threshold`.

Esses diagnósticos não substituem a regra oficial adotada nesta execução, mas fortalecem
a rastreabilidade metodológica da etapa.

Este notebook define a estrutura fundamental de episódios críticos utilizada no restante
do pipeline.

In [1]:
# ============================================================
# 04_window_5min_episodes.ipynb
# Detecção de episódios críticos (μ + 2σ) na série 5-min
# Métrica oficial: taxa de falha (fail_rate = n_failed / n_events)
# Pipeline PPCOMP_DM (Google Cluster Trace)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd
from IPython.display import display

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# Mount Google Drive (seguro)
if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull)...")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso: não foi possível atualizar via git pull:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs
ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH =", REPORTS_PATH)

def log(msg: str) -> None:
    print(f"[04_detect_episodes] {msg}")

# -----------------------------
# 1) Metadados do NB03
# -----------------------------
NB03_SUMMARY_FILE = REPORTS_PATH / "03_window_5min_base_summary.json"

nb03_summary = {}
if NB03_SUMMARY_FILE.exists():
    nb03_summary = json.loads(NB03_SUMMARY_FILE.read_text(encoding="utf-8"))
    log(f"Resumo do NB03 carregado: {NB03_SUMMARY_FILE}")
else:
    log("Resumo do NB03 não encontrado; seguindo sem metadados adicionais.")

SCENARIO_LABEL = nb03_summary.get("scenario_label", "unknown")
REMOVE_HOUR_ZERO = nb03_summary.get("remove_hour_zero", None)

# -----------------------------
# 2) Leitura da série contínua
# -----------------------------
SERIES_FILE = FEATURES_PATH / "window_5min_series.parquet"
assert SERIES_FILE.exists(), f"Arquivo não encontrado: {SERIES_FILE}"

df = pd.read_parquet(SERIES_FILE)
log(f"Shape entrada: {df.shape}")

required_cols = {"bucket_id", "bucket_start_us", "n_failed", "n_events"}
missing = required_cols - set(df.columns)
assert not missing, f"Colunas ausentes: {missing}"

df = df.sort_values("bucket_id").reset_index(drop=True)

assert df["bucket_id"].is_monotonic_increasing, "bucket_id não está ordenado."
assert df["bucket_id"].is_unique, "Foram encontrados bucket_id duplicados."

bucket_diff = df["bucket_id"].diff().dropna()
n_internal_gaps_gt1 = int((bucket_diff > 1).sum())
assert n_internal_gaps_gt1 == 0, (
    "A série contínua deveria estar sem lacunas internas após o NB03."
)

series_rows = int(len(df))
bucket_min = int(df["bucket_id"].min())
bucket_max = int(df["bucket_id"].max())

log(f"Série contínua: rows={series_rows}, bucket range={bucket_min}..{bucket_max}")

# -----------------------------
# 3) Métrica oficial: fail_rate
# -----------------------------
zero_event_windows = int((df["n_events"] == 0).sum())
nonzero_event_windows = int((df["n_events"] > 0).sum())

den = df["n_events"].replace(0, np.nan)
df["fail_rate"] = (df["n_failed"] / den).fillna(0.0).astype("float32")

log(f"Janelas com n_events == 0: {zero_event_windows}")
log(f"Janelas com n_events  > 0: {nonzero_event_windows}")

# -----------------------------
# 4) Estatística global oficial
# -----------------------------
mu = float(df["fail_rate"].mean())
sigma = float(df["fail_rate"].std(ddof=0))
threshold = float(mu + 2 * sigma)

df["is_critical"] = (df["fail_rate"] >= threshold).astype("int8")
critical_windows = int(df["is_critical"].sum())

log(f"mu={mu:.8f}")
log(f"sigma={sigma:.8f}")
log(f"threshold={threshold:.8f}")
log(f"Total janelas críticas: {critical_windows}")

# -----------------------------
# 5) Diagnóstico do bucket 0
# -----------------------------
bucket0_present = bool((df["bucket_id"] == 0).any())

bucket0_summary = None
if bucket0_present:
    row0 = df.loc[df["bucket_id"] == 0].iloc[0]
    bucket0_summary = {
        "bucket_id": int(row0["bucket_id"]),
        "bucket_start_us": int(row0["bucket_start_us"]),
        "n_events": int(row0["n_events"]),
        "n_failed": int(row0["n_failed"]),
        "fail_rate": float(row0["fail_rate"]),
        "is_critical": int(row0["is_critical"]),
    }
    log(f"Bucket 0 summary: {bucket0_summary}")
else:
    log("Bucket 0 não encontrado na série contínua.")

# -----------------------------
# 6) Diagnóstico de sensibilidade (sem alterar regra oficial)
#    6.1 Excluindo bucket 0
# -----------------------------
diag_without_bucket0 = None
if bucket0_present and len(df) > 1:
    df_wo_b0 = df[df["bucket_id"] != 0].copy()
    mu_wo_b0 = float(df_wo_b0["fail_rate"].mean())
    sigma_wo_b0 = float(df_wo_b0["fail_rate"].std(ddof=0))
    threshold_wo_b0 = float(mu_wo_b0 + 2 * sigma_wo_b0)
    critical_windows_wo_b0 = int((df_wo_b0["fail_rate"] >= threshold_wo_b0).sum())

    diag_without_bucket0 = {
        "rows": int(len(df_wo_b0)),
        "mu": mu_wo_b0,
        "sigma": sigma_wo_b0,
        "threshold": threshold_wo_b0,
        "critical_windows": critical_windows_wo_b0,
    }

    log(
        "Diag sem bucket 0 -> "
        f"mu={mu_wo_b0:.8f}, sigma={sigma_wo_b0:.8f}, "
        f"threshold={threshold_wo_b0:.8f}, critical_windows={critical_windows_wo_b0}"
    )

# -----------------------------
# 7) Consolidação de episódios
# -----------------------------
episodes = []
in_episode = False
start_idx = None

for i, flag in enumerate(df["is_critical"].to_numpy()):
    if flag == 1 and not in_episode:
        in_episode = True
        start_idx = i
    elif flag == 0 and in_episode:
        episodes.append((start_idx, i - 1))
        in_episode = False

if in_episode:
    episodes.append((start_idx, len(df) - 1))

log(f"Episódios detectados: {len(episodes)}")

# -----------------------------
# 8) Métricas por episódio
# -----------------------------
WINDOW_MINUTES = 5
rows = []

for ep_id, (s, e) in enumerate(episodes, start=1):
    seg = df.iloc[s:e+1]

    rows.append({
        "episode_id": ep_id,
        "start_bucket": int(seg["bucket_id"].iloc[0]),
        "end_bucket": int(seg["bucket_id"].iloc[-1]),
        "start_bucket_start_us": int(seg["bucket_start_us"].iloc[0]),
        "end_bucket_start_us": int(seg["bucket_start_us"].iloc[-1]),
        "duration_windows": int(len(seg)),
        "duration_minutes": int(len(seg) * WINDOW_MINUTES),

        # oficial (taxa)
        "max_fail_rate": float(seg["fail_rate"].max()),
        "mean_fail_rate": float(seg["fail_rate"].mean()),

        # diagnóstico adicional (volume)
        "max_failed": int(seg["n_failed"].max()),
        "mean_failed": float(seg["n_failed"].mean()),
        "sum_failed": int(seg["n_failed"].sum()),
        "max_events": int(seg["n_events"].max()),
        "mean_events": float(seg["n_events"].mean()),

        # parâmetros globais oficiais
        "threshold": float(threshold),
        "mu": float(mu),
        "sigma": float(sigma),
    })

episodes_df = pd.DataFrame(rows)

log(f"Episodes DF shape: {episodes_df.shape}")

# -----------------------------
# 9) Estatísticas de duração dos episódios
# -----------------------------
if len(episodes_df) > 0:
    one_window_episodes = int((episodes_df["duration_windows"] == 1).sum())
    duration_mean = float(episodes_df["duration_windows"].mean())
    duration_median = float(episodes_df["duration_windows"].median())
    duration_max = int(episodes_df["duration_windows"].max())
else:
    one_window_episodes = 0
    duration_mean = 0.0
    duration_median = 0.0
    duration_max = 0

# -----------------------------
# 10) Persistência
# -----------------------------
OUT_FILE = FEATURES_PATH / "episodes_detected.parquet"
OUT_SCENARIO_FILE = FEATURES_PATH / f"episodes_detected_{SCENARIO_LABEL}.parquet"

episodes_df.to_parquet(OUT_FILE, compression="snappy", index=False)
episodes_df.to_parquet(OUT_SCENARIO_FILE, compression="snappy", index=False)

log(f"Episodes salvos (canônico): {OUT_FILE}")
log(f"Episodes salvos (cenário): {OUT_SCENARIO_FILE}")

# Opcional e útil para rastreabilidade: série enriquecida com fail_rate/is_critical
SERIES_SCORED_FILE = FEATURES_PATH / "window_5min_series_scored.parquet"
SERIES_SCORED_SCENARIO_FILE = FEATURES_PATH / f"window_5min_series_scored_{SCENARIO_LABEL}.parquet"

df.to_parquet(SERIES_SCORED_FILE, compression="snappy", index=False)
df.to_parquet(SERIES_SCORED_SCENARIO_FILE, compression="snappy", index=False)

log(f"Série enriquecida salva (canônica): {SERIES_SCORED_FILE}")
log(f"Série enriquecida salva (cenário): {SERIES_SCORED_SCENARIO_FILE}")

# -----------------------------
# 11) Summary JSON
# -----------------------------
summary = {
    "metric": "fail_rate",
    "scenario_label": SCENARIO_LABEL,
    "remove_hour_zero": REMOVE_HOUR_ZERO,

    "input_series_file": str(SERIES_FILE),
    "series_rows": int(series_rows),
    "bucket_id_min": int(bucket_min),
    "bucket_id_max": int(bucket_max),

    "zero_event_windows": int(zero_event_windows),
    "nonzero_event_windows": int(nonzero_event_windows),

    "mu": float(mu),
    "sigma": float(sigma),
    "threshold": float(threshold),
    "critical_windows": int(critical_windows),
    "episodes_detected": int(len(episodes_df)),

    "bucket0_present": bool(bucket0_present),
    "bucket0_summary": bucket0_summary,
    "diag_without_bucket0": diag_without_bucket0,

    "one_window_episodes": int(one_window_episodes),
    "duration_windows_mean": float(duration_mean),
    "duration_windows_median": float(duration_median),
    "duration_windows_max": int(duration_max),

    "output_episodes_file_canonical": str(OUT_FILE),
    "output_episodes_file_scenario": str(OUT_SCENARIO_FILE),
    "output_series_scored_file_canonical": str(SERIES_SCORED_FILE),
    "output_series_scored_file_scenario": str(SERIES_SCORED_SCENARIO_FILE),
}

SUMMARY_FILE = REPORTS_PATH / "04_detect_episodes_summary.json"
SUMMARY_SCENARIO_FILE = REPORTS_PATH / f"04_detect_episodes_summary_{SCENARIO_LABEL}.json"

SUMMARY_FILE.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")
SUMMARY_SCENARIO_FILE.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")

log(f"Resumo salvo (canônico): {SUMMARY_FILE}")
log(f"Resumo salvo (cenário): {SUMMARY_SCENARIO_FILE}")

# -----------------------------
# 12) Visualização rápida
# -----------------------------
print("\n=== RESUMO RÁPIDO ===")
print(f"Scenario label            : {SCENARIO_LABEL}")
print(f"Remove hour 0             : {REMOVE_HOUR_ZERO}")
print(f"Series rows               : {series_rows}")
print(f"Bucket range              : {bucket_min}..{bucket_max}")
print(f"Zero-event windows        : {zero_event_windows}")
print(f"mu                        : {mu:.8f}")
print(f"sigma                     : {sigma:.8f}")
print(f"threshold                 : {threshold:.8f}")
print(f"Critical windows          : {critical_windows}")
print(f"Episodes detected         : {len(episodes_df)}")
print(f"Bucket 0 present          : {bucket0_present}")
print(f"One-window episodes       : {one_window_episodes}")
print(f"Duration mean (windows)   : {duration_mean:.4f}")
print(f"Duration median (windows) : {duration_median:.4f}")
print(f"Duration max (windows)    : {duration_max}")

if bucket0_summary is not None:
    print("\nBucket 0 summary:")
    print(bucket0_summary)

if diag_without_bucket0 is not None:
    print("\nDiag sem bucket 0:")
    print(diag_without_bucket0)

print("\n=== HEAD EPISODES ===")
display(episodes_df.head())

print("\n=== HEAD SERIES SCORED ===")
display(df.head())

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull)...
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH = /content/drive/MyDrive/Mestrado/04-reports
[04_detect_episodes] Resumo do NB03 carregado: /content/drive/MyDrive/Mestrado/04-reports/03_window_5min_base_summary.json
[04_detect_episodes] Shape entrada: (8930, 18)
[04_detect_episodes] Série contínua: rows=8930, bucket range=0..8929
[04_detect_episodes] Janelas com n_events == 0: 2
[04_detect_episodes] Janelas com n_events  > 0: 8928
[04_detect_episodes] mu=0.19657958
[04_detect_episodes] sigma=0.14091089
[04_detect_episodes] threshold=0.47840136
[04_detect_episodes] Total janelas críticas: 390
[04_detect_episodes] Bucket 0 summary: {'bucket_id': 0, 'bucket_start_us': 0, 'n_events': 56452, 'n_failed': 37082, 'fail_rate': 0.6568766236305237, 'is_critical': 1}
[04_detect_episodes] Diag sem bucket 0 -> mu=0.19652800, sigm

,episode_id,start_bucket,end_bucket,start_bucket_start_us,end_bucket_start_us,duration_windows,duration_minutes,max_fail_rate,mean_fail_rate,max_failed,mean_failed,sum_failed,max_events,mean_events,threshold,mu,sigma
0,1,0,0,0,0,1,5,0.656877,0.656877,37082,37082.0,37082,56452,56452.0,0.478401,0.19658,0.140911
1,2,666,666,199800000000,199800000000,1,5,0.500000,0.500000,7,7.0,7,14,14.0,0.478401,0.19658,0.140911
2,3,967,967,290100000000,290100000000,1,5,0.756757,0.756757,84,84.0,84,111,111.0,0.478401,0.19658,0.140911
3,4,1220,1220,366000000000,366000000000,1,5,0.625000,0.625000,10,10.0,10,16,16.0,0.478401,0.19658,0.140911
4,5,1391,1391,417300000000,417300000000,1,5,0.750000,0.750000,6,6.0,6,8,8.0,0.478401,0.19658,0.140911



=== HEAD SERIES SCORED ===


,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,req_mem_presence_rate,event_FAIL_count,event_SCHEDULE_count,event_FINISH_count,event_ENABLE_count,event_LOST_count,event_EVICT_count,event_KILL_count,fail_rate,is_critical
0,0,0,56452,37082,43394,1232,166.682739,0.018434,0.009607,0.986289,0.986289,37082,154,600,18616,0,0,0,0.656877,1
1,1,300000000,0,0,0,0,NaN,NaN,NaN,0.000000,0.000000,0,0,0,0,0,0,0,0.000000,0
2,2,600000000,28,9,28,19,208.464286,0.008100,0.016303,1.000000,1.000000,9,0,7,7,5,0,0,0.321429,0
3,3,900000000,32,2,32,18,135.656250,0.009495,0.003928,1.000000,1.000000,2,0,19,4,6,0,1,0.062500,0
4,4,1200000000,29,6,28,20,233.068966,0.012932,0.003717,1.000000,1.000000,6,0,6,6,11,0,0,0.206897,0


## 10. Conclusão da Etapa

A etapa de detecção de episódios críticos foi concluída com sucesso, preservando a regra
oficial do pipeline baseada em taxa de falha por janela:

`fail_rate = n_failed / n_events`

com critério global de criticidade dado por:

`threshold = μ + 2σ`

Nesta execução, a série contínua analisada apresentou:
- 8.930 janelas no total;
- intervalo de buckets entre 0 e 8929;
- 2 janelas com `n_events == 0`;
- 8.928 janelas com `n_events > 0`.

A aplicação da regra oficial resultou em:
- `μ = 0.19657958`
- `σ = 0.14091089`
- `threshold = 0.47840136`
- 390 janelas críticas
- 285 episódios detectados

Com isso, o NB04 produziu a segmentação crítica oficial que será utilizada nas etapas
posteriores do pipeline.

## 11. Achados Experimentais da Execução Atual

A execução atual trouxe um resultado especialmente importante: a nova série iniciando em
`bucket_id = 0` alterou levemente os parâmetros estatísticos globais e o número de
janelas críticas, sem alterar o número total de episódios detectados.

### 11.1 Estrutura da série analisada

A série contínua utilizada nesta execução apresentou:
- `series_rows = 8930`
- `bucket_id_min = 0`
- `bucket_id_max = 8929`
- `zero_event_windows = 2`
- `nonzero_event_windows = 8928`

Esses valores são coerentes com a saída do NB03, confirmando que a detecção de episódios
foi realizada sobre a nova série contínua derivada do cenário `keep_hour0`.

### 11.2 Parâmetros estatísticos oficiais

A taxa de falha foi calculada em todas as janelas, com tratamento de borda:
- para `n_events = 0`, definiu-se `fail_rate = 0.0`.

A partir da distribuição global de `fail_rate`, foram obtidos:
- média global `μ = 0.1965795755`
- desvio-padrão global `σ = 0.1409108937`
- limiar oficial `threshold = 0.4784013629`

A classificação crítica utilizou a regra:
- uma janela é crítica quando `fail_rate >= threshold`.

Com essa regra, foram identificadas:
- 390 janelas críticas
- 285 episódios críticos

### 11.3 Bucket inicial da série

O bucket inicial (`bucket_id = 0`) esteve presente explicitamente na série e apresentou
os seguintes valores:
- `bucket_start_us = 0`
- `n_events = 56452`
- `n_failed = 37082`
- `fail_rate = 0.65687662`
- `is_critical = 1`

Ou seja, o bucket 0 foi classificado como crítico nesta execução.

Esse resultado confirma que o início da série não apenas existe formalmente, mas também
tem impacto real sobre a estatística global e sobre a segmentação de criticidade.

### 11.4 Diagnóstico complementar sem o bucket 0

Como análise complementar, foi calculado um diagnóstico removendo apenas o bucket 0,
sem alterar a regra oficial do notebook.

Nesse cenário alternativo, os parâmetros ficaram:
- `rows = 8929`
- `μ = 0.19652800`
- `σ = 0.14083412`
- `threshold = 0.47819625`
- `critical_windows = 394`

Comparando com a execução oficial:
- com bucket 0: 390 janelas críticas
- sem bucket 0: 394 janelas críticas

Portanto, a presença do bucket 0:
- aumenta ligeiramente `μ` e `σ`;
- aumenta levemente o limiar;
- reduz o número total de janelas críticas de 394 para 390.

Contudo, o número de episódios detectados permaneceu em 285, indicando que o efeito da
preservação do bucket 0, nesta etapa, concentrou-se sobretudo na contagem de janelas
críticas e não na quantidade total de episódios.

Esse achado é importante porque fortalece a argumentação metodológica: a preservação da
primeira janela altera a estatística global de forma mensurável, mas não produz ruptura
substancial na segmentação episódica.

### 11.5 Perfil dos episódios detectados

Os episódios detectados apresentaram as seguintes características:
- `episodes_detected = 285`
- `one_window_episodes = 227`
- `duration_windows_mean = 1.3684`
- `duration_windows_median = 1.0`
- `duration_windows_max = 7`

Esses resultados mostram que:
- a maioria dos episódios é curta;
- a mediana de duração é de uma única janela;
- há poucos episódios mais longos, com duração máxima de 7 janelas.

Esse perfil é compatível com uma série em granularidade de 5 minutos e com uma regra de
criticidade baseada em desvios significativos da taxa de falha por janela.

## 12. Importância para o Pipeline Analítico

O NB04 estabelece a estrutura crítica oficial do pipeline:
- identifica janelas críticas;
- consolida essas janelas em episódios;
- produz um artefato persistido contendo a segmentação episódica;
- registra a série enriquecida com `fail_rate` e `is_critical`.

Essa etapa é fundamental porque serve de referência para:
- engenharia de atributos no NB05;
- definição de estados no NB06;
- formulação do problema supervisionado nas etapas posteriores.

Além disso, a execução atual melhora a rastreabilidade científica do pipeline ao mostrar,
de forma explícita, como a preservação do início da série afeta os parâmetros globais e a
quantidade de janelas críticas, sem alterar a quantidade total de episódios detectados.

## 13. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos:
- `episodes_detected.parquet`
- `episodes_detected_keep_hour0.parquet`
- `window_5min_series_scored.parquet`
- `window_5min_series_scored_keep_hour0.parquet`
- `04_detect_episodes_summary.json`
- `04_detect_episodes_summary_keep_hour0.json`

Esses arquivos passam a constituir a referência oficial da etapa de detecção de episódios
para o cenário `keep_hour0`.

## 14. Encaminhamento do Pipeline

Com os episódios críticos devidamente detectados e persistidos, o pipeline segue para:
- NB05 — Engenharia de atributos;
- NB06 — Derivação de estados e formulação supervisionada;
- etapas posteriores de modelagem e avaliação.

A partir deste ponto, as demais etapas passam a utilizar como base a segmentação crítica
oficial produzida nesta execução.